Add motion planning pages: end effector frames, waypoints, pose clouds, plan verification (#5098)
viamrobotics/docs: docs/motion-planning/frame-system/end-effector-frames.md

---
linkTitle: "End effector frames"
title: "End effector frames"
weight: 5
layout: "docs"
type: "docs"
description: "How the motion service moves a named frame to a target pose, and how the frame attribute and WorldState transforms control which point on the robot that frame represents."

> ---

Reviewer: Keeping a few general questions here as I review this:
- What can we expect readers to already be familiar with when they read this page?
- Do we need to summarize what a pose is? Should there be a reference section, perhaps in the overview that we link readers to for more details?
- Same question for world state.

Reviewer: Yeah, I think we need to define pose and frame clearly in a place that we can reference elsewhere. Currently thinking the overview page might be the best place for that, but I'll continue to assess as I review.

> When you ask the motion service to move an arm, you do not move "the arm." You

Reviewer: This personifies the motion service. That's the wrong tone for docs. As an alternative, you might say

To move an arm using the Viam motion service, you must specify a desired pose for a specific frame defined within the frame system. The motion planner solves for joint angles that place that frame's origin in the destination pose....

move a _frame_. The destination you pass is a pose for a named frame in the
frame system, and the planner solves for joint angles that place that frame's

> origin at the destination. Understanding which frame you are moving, and where

Reviewer: Rather than the sentence beginning, "Understanding..." we need to make the abstract description that precedes it concrete -- something like...the frame you specify is typically the frame of the end effector for the arm...

that frame sits on the physical robot, is the difference between the tool tip
landing on the target and the arm's mounting flange landing there instead.

> This page explains how the motion service targets frames, how the `frame`

Reviewer: I think we can more clearly explain this than saying "targets frames". What does that mean?

attribute defines an end effector frame, and how a `WorldState` transform lets

> you add or reposition that frame for a single request.

Reviewer: Can you explain what this page is without using so many terms that might not be familiar to the reader before they read the page?

## Move targets a frame, not a component

> The `Move` request takes a component name and a destination. The component name

Reviewer: We need to be precise. `Move` is a method in the motion service API. Can we be more precise about what the destination is exactly?

> is not treated as a piece of hardware to actuate directly; it is resolved to a

Reviewer: Typically, for docs, we don't want to say what something is not...So this sentence can be The component name resolves to a defined frame in the frame system.

> frame in the frame system. The planner then searches for an arm configuration

Reviewer: Problematic to use the word, "configuration" here given how much we talk about configuration generally with Viam.

---
linkTitle: "End effector frames"
title: "End effector frames"
weight: 5
layout: "docs"
type: "docs"
description: "How Move places a named frame at a target pose, and how the frame attribute and WorldState transforms set which point on the robot a frame represents."
---

When you move an arm with the motion service, you move a _frame_ from its current
pose to a destination pose. That frame is called the **end effector frame**, also called
the tool control point (TCP) or the manipulation control frame. Where the end effector
frame sits depends on the arm, the tool, and the task, and that frame can change partway

> through a job. This page explains why the end effector frame exists, how `Move` resolves it,

Reviewer: We're introducing Move for the first time on this page and need to explain what it is.

and how you define and visualize it with Viam.

First, look at where the end effector frame could be defined for two example use cases:

> {{<imgproc src="/motion-planning/frame-system/end-effector-types.svg" declaredimensions=true alt="Two arms drawn schematically. On the left, a two-finger gripper with its end effector frame between the fingertips. On the right, a drill whose end effector frame sits at the bit tip, offset from the arm frame in both x and z. Both panels also show small coordinate axes for the world frame at the base and the arm frame at the flange." style="max-width:820px" class="aligncenter">}}

Reviewer: The vectors in these figures are mostly easy to understand, but the length of the vector leaves us with just the arrow head for most of them. That is so distracting that I think we need to fix it.

Author: extended, it looks much better now

The end effector frame is hardware and task dependent. For a two-finger gripper, the

> end effector frame sits between the fingertips, a short offset below the arm's frame. For

Reviewer: ```suggestion
end effector frame typically sits between the fingertips, a short offset below the arm's frame. For
```

Reviewer: Also, we either need to explain what we mean by the "arm's frame" and direct the reader to the figure.

a drill, it sits at the bit tip, offset from the arm's frame both forward and down,

> and the orientation may be flipped, so a positive z would mean to drill deeper.

Reviewer: In what cases would it not be flipped? Is it even flipped? Isn't z always positive moving away from the end of the arm?

Author: It typically is z+ out of the last joint for Cobots and industrial arms (they usually use DH), but not for humanoids, where the frame after joint is more arbitrary and typically defined by the zero pose. I can update the examples to be more cobot related

When you call `Move`, you have the option to specify which frame you are moving, and
using the arm frame is always an option. By creating an end effector frame at the
correct location, a `Move` call becomes easier and better represents the task being

> performed.

Reviewer: This entire paragraph should be more directive -- "do it this way"

> Now look at how the motion service describes the move. The snippets on this page

Reviewer: This is worded weirdly. I think we're showing an example next, right? So, we'd write this something like, "See below for an example of..."

assume a connected motion service client, `motion_service` in Python and
`motionService` in Go. For how to set this up in more detail, see
[Access the motion service in your code](/reference/services/motion/#access-the-motion-service-in-your-code).

```python
from viam.proto.common import Pose, PoseInFrame

# A pose in world frame facing Z Down
target_pose = PoseInFrame(
    reference_frame="world",
    pose=Pose(x=400, y=0, z=300, o_x=0, o_y=0, o_z=-1, theta=0),
)

# Move the Gripper Frame to the target_pose
await motion_service.move(
    component_name="my-gripper",
    destination=target_pose,
)
```

You pass two parameters to `Move`: a string `component_name` and a `destination`
pose.

## Which frame Move resolves

The `destination` you pass is a target pose for the resolved frame, and the planner

> tries to find a collision-free motion that moves the frame to the destination.

Reviewer: Do we need a bit more here about move all joints through a series of way points to achieve the destination position?

The `component_name` can be a component name or a frame name. The motion service
looks for a frame with that name in the frame system. When you pass a component name

> such as `"my-arm"` or `"my-gripper"`, Viam supplies a frame for it behind the scenes:

Reviewer: I think we should just specify that you should use a component name and that Viam uses the defined frame for that component. In general, we need to be directive. Where there is a real "and also", we should state the preferred approach first and then mention the alternative. Here that will just be confusing.

Author: I would be happy to change it, but I think it would be worth using it to teach frames here. I would want a user to understand clearly the thing they pass in is a frame name. We shield users from frames, albeit a little inconsistently, and to do more useful stuff with the platform they will need to get more comfortable with them.

- **An arm**: its kinematics file (URDF or Viam JSON) defines a single output frame at the
  terminal link of the kinematic chain, commonly called the tool flange. That output frame

>   is the arm's built-in end effector frame.

Reviewer: Here are we just clarifying that for any arm that Viam supports it maintains a frame at the tool flange? We discuss that above around where the graphics are. Here, in part because it's in a bullet, it looks like we're introducing something new. I think all we're really doing here is getting more specific about the gripper frame and trying to distinguish from the arm frame? I think we can simply remind the reader that the arm frame is rooted at the tool flange and refer them to the graphic above. Don't do this in bullets.

- **Another component**, such as a gripper, defines a static frame through its `frame`
  attribute, which you set on the component's configuration card.

> Here is a gripper `frame` attribute from the [Arm with gripper and wrist camera](/motion-planning/frame-system/arm-gripper-camera/) example, whose origin sits 120 mm past the flange:

Reviewer: "origin" is a new term here. We either need to define it or use another word that readers will already understand.

```json
{
  "parent": "my-arm",
  "translation": { "x": 0, "y": 0, "z": 120 },
  "orientation": {
    "type": "ov_degrees",
    "value": { "x": 0, "y": 0, "z": 1, "th": 0 }
  }
}
```

## Why the frame you name matters

A `Move` call that uses the arm frame produces different motion than one that

> uses the drill-tip frame. The arm's frame sits at the flange, and the drill-tip

Reviewer: ```suggestion
uses the end-effector frame. The arm's frame sits at the flange, and the end-effector
```

frame sits at the bit tip, so the same `target_pose` produces different motions.
In the example below, the target pose is defined on the face of a box. If you use
the drill-tip frame, the bit moves to the box face. If you use the arm frame, the
drill drives into the box, because the planner tries to move the arm frame to the

> target pose.

Reviewer: This is such a good example.

{{<imgproc src="/motion-planning/frame-system/arm-vs-gripper-frame.svg" declaredimensions=true alt="Two diagrams of the same target pose. On the left, the move targets the arm frame, so the flange lands on the target and the drill bit overshoots past it. On the right, the move targets the drill-tip frame, so the bit lands on the target. Both panels show the world frame, the arm frame, and the labeled target pose." style="max-width:760px" class="aligncenter">}}

## Define an end effector frame in code

The `frame` attribute described above works well when the end effector frame is static.

> When the end effector frame moves during a task, or when defining it in code suits

Reviewer: So this reads as though I would always need to use the WorldState transform, because at this point, I could interpret my end effector frame as moving any time I call Move. Let's revise this to remove that potential misread.

the job better, add a new frame to the frame system with a `WorldState` transform.

> A great example is a grasped part. Prior to pickup, the end effector frame is located

Reviewer: ```suggestion
One example is when your end effector grasps an object. Prior to pickup, the end effector frame is located
```

"great example" is hyperbole we'd like to avoid.

with the gripper. After pickup, the end effector frame may need to move to the part for
an assembly step. The grasped part may not have been in the machine configuration,
so you add its frame to the frame system at request time with a `WorldState` transform,
attached to the gripper so it moves with the grasp:

{{<imgproc src="/motion-planning/frame-system/held-object-frame.svg" declaredimensions=true alt="A cobot arm whose gripper holds a bent, L-shaped object. The arm frame sits at the flange and the gripper frame at the jaws. An object-tip frame sits at the far end of the object, offset from the gripper in multiple axes and rotated to point along it. Each frame shows labeled x, y, and z axes." style="max-width:600px" class="aligncenter">}}

To extend that frame system for
one motion, pass a
[`WorldState`](https://python.viam.dev/autoapi/viam/proto/common/index.html#viam.proto.common.WorldState)
to `Move` with extra frames. The steps differ slightly by language, but the shape is
the same:

> 1. Define the translation and orientation from the parent frame to the origin of the

Reviewer: We need to say something like, in the example above, the parent frame is the gripper.

   new frame. In Python, this is a `PoseInFrame`; in Go, it is the pose you pass to
   `NewLinkInFrame`. The parent is usually an arm or gripper component name, such as
   `"my-gripper"` for a grasped object.
2. Build a transform from that pose and add it to the `WorldState`.

The `WorldState` applies to a single `Move` call. A later call needs its own
`WorldState` to keep using the transform.

{{< tabs >}}
{{% tab name="Python" %}}

```python
from viam.proto.common import (
    Transform, PoseInFrame, Pose, WorldState
)

> # A frame named "object-tip", 100 mm beyond the gripper, attached to it.

Reviewer: The box for the code seems to be narrower than the page on many widths. Use Claude to figure out what you need to add as attributes or styling on the code tabs to use available width and avoid the need to scroll right. I think we can also be better about comment placement to avoid the need to scroll.

Author: This seems to be a preview issue only, i've seen it a few times, claude says we can fix it, but there was some worry that the fix could negatively affect the real deployment, will look into more outside of this pr

await motion_service.move(
    component_name="object-tip",
    destination=target_pose,
    world_state=world_state,
)
```

{{% /tab %}}
{{% tab name="Go" %}}

```go
// A frame named "object-tip", 100 mm beyond the gripper, attached to it.
objectTipLink := referenceframe.NewLinkInFrame(
    "my-gripper", // the parent: the object moves with the gripper
    // 70 mm across and 130 mm down from the gripper, turned so z points along the bent object.
    spatialmath.NewPose(
        r3.Vector{X: 70, Y: 0, Z: 130},
        &spatialmath.OrientationVectorDegrees{OX: 1, OY: 0, OZ: 0, Theta: 90},
    ),
    "object-tip", // the name of the new frame
    nil,          // optional collision geometry; nil means a pure coordinate frame
)

worldState, err := referenceframe.NewWorldState(
    nil, []*referenceframe.LinkInFrame{objectTipLink})
if err != nil {
    logger.Fatal(err)
}

// Move can now place "object-tip" at the destination.
_, err = motionService.Move(ctx, motion.MoveReq{
    ComponentName: "object-tip",
    Destination:   targetPose, // your goal *referenceframe.PoseInFrame
    WorldState:    worldState,
})
```

{{% /tab %}}
{{< /tabs >}}

> ### Attach a geometry to a transform

Reviewer: We're introducing geometry here. We need to define it and differentiate it from the frame.

> A transform can also carry a collision geometry, a shape the planner avoids while it

Reviewer: We kind-of define it in this sentence, but let's be explicit and differentiate from frame.

        reference_frame="my-gripper",
        pose=Pose(x=70, y=0, z=130, o_x=1, o_y=0, o_z=0, theta=90),
    ),
    physical_object=Geometry(  # the collision geometry for this frame
        center=Pose(x=0, y=0, z=-50, o_x=0, o_y=0, o_z=1, theta=0),  # centered on the grasped object
        box=RectangularPrism(dims_mm=Vector3(x=30, y=30, z=100)),
        label="object-tip",
    ),
)
```

{{% /tab %}}
{{% tab name="Go" %}}

```go
// Make a new 30 x 30 x 100 mm box covering the grasped object.
geometry, err := spatialmath.NewBox(
    spatialmath.NewPoseFromPoint(r3.Vector{X: 0, Y: 0, Z: -50}), // centered on the grasped object
    r3.Vector{X: 30, Y: 30, Z: 100},                             // full dimensions in mm
    "object-tip",
)
if err != nil {
    logger.Fatal(err)
}

objectTipLink := referenceframe.NewLinkInFrame(
    "my-gripper",
    spatialmath.NewPose(
        r3.Vector{X: 70, Y: 0, Z: 130},
        &spatialmath.OrientationVectorDegrees{OX: 1, OY: 0, OZ: 0, Theta: 90},
    ),
    "object-tip",
    geometry, // the planner avoids collisions with this geometry
)
```

{{% /tab %}}
{{< /tabs >}}

## WorldState versus the world state store service

> The names are similar, but `WorldState` and the world state store service are

Reviewer 2: I think the goal is to populate motion service requests with items defined in the world state store service. Not sure if that task without a completion date warrants changing the narrative here. But just wanted to be transparent that the two ought to be more similar than they currently are.

The documentation though is accurate w.r.t how things behave today.

Author: Once that goes through I will need to update a bunch then, have similar distinctions in the new Visualization Section

different things, and only one affects a `Move` call.

- `WorldState` is the argument you pass to a single `Move` call. The obstacles
  and transforms in it exist only for that request: the planner uses them to
  plan this motion, and they are gone when the call returns. The transform in
  the previous section is a `WorldState` transform, so it shapes the plan.
- The world state store service is a separate service that holds transforms for

>   client-side visualization in the [3D scene](/motion-planning/3d-scene/).

Reviewer: This is the first mention of world state store service so readers don't even know about it before now. Where will they encounter the service? We need to call that out and make this more of a "heads up at some point you're going to come across the world state store service. Here's what that is and how it differs from WorldState.

To change where the arm goes, put the transform in the `WorldState` you pass to
`Move`, as shown above. To render a custom visual that leaves planning unchanged,
publish it to the world state store service instead.

## Visualize the end effector frame

> Static frames render in the 3D scene automatically. Open the **3D SCENE** tab on

Reviewer: What's a static frame? We've kind-of introduced that, but should be more explicit here.
